Kernel handshaking pattern proposal (#66)
jupyter/enhancement-proposals: jupyter-handshaking/jupyter-handshaking.md

# Kernel Handshaking pattern

## Problem

The current implementation of Jupyter client makes it responsible for finding available ports and pass them to a new starting kernel. The issue is that a new process can start using one of these ports before the kernel has started, resulting in a ZMQError when the kernel starts. This is even more problematic when spawning a lot of kernels in a short laps of time, because the client may find available ports that have already been assigned to another kernel.

A workaround has been implemented for the latter case, but it does not solve the former one.

## Proposed Enhancement

We propose to implement a handshaking pattern: the client lets the kernel find free ports and communicate them back via a dedicated socket. It then connects to the kernel. More formely:

> - When the client starts, it opens a dedicated socket A for receiving connection information from kernels (channel ports).

Reviewer: What is the lifetime of this dedicated socket?  Is it per-kernel launch or once per "application".  I suppose this could be considered an implementation detail, but I think the JEP should minimally state that.  

However, so that different implementations _can_ succeed, I think we should minimally include the kernel's ID in the response information sent to the kernel and require the kernel to include it in the connection information it returns.  This would then allow implementations to use a "dedicated socket" that spans multiple kernel launches, should they choose to do so.

Reviewer 2: I would consider it an implementation detail, but it's worth specifying that:

1. the kernel should not expect the socket to exist after it has sent its registration (i.e. it _can_ be closed).
2. a recommended implementation for a multi-kernel client (i.e. jupyter-server) is to have a single long-lived registration socket.
3. the kernel should close its registration socket after completion.

> - When launching a new kernel, the client passes its address and the port of this socket to the kernel.

Reviewer 2: ```suggestion
- When launching a new kernel, the client passes the connection information for this socket to the kernel.
```

> We propose to implement a handshaking pattern: the client lets the kernel find free ports and communicate them back via a dedicated socket. It then connects to the kernel. More formely:
> 
> - When the client starts, it opens a dedicated socket A for receiving connection information from kernels (channel ports).
> - When launching a new kernel, the client passes its address and the port of this socket to the kernel.
> - The kernel starts, find free ports to bind the shell, control, stdin, heartbeat and iopub sockets. It then connect to the A socket and send the connection information to the client.
> - Upon reception of the connection information, the client connects to the kernel.
> 
> The way the client passes its address and the port of the listening socket to the kernel should be similar to that of passing the ports of the kernel socket in the current implementation: a connection file that can be read by local kernels or sent over the network for remote kernels (although this requires a custom kernel provisioner or "nanny"). This connection file should also contain the signature scheme and the key.

Reviewer: I think it would be helpful to apply a logical _name_ to the socket...
```suggestion
We propose to implement a handshaking pattern: the client lets the kernel find free ports and communicate them back via a dedicated _response socket_. It then connects to the kernel. More formally:

- When the client starts, it opens a dedicated _respsonse socket_ for receiving connection information from kernels (channel ports).
- When launching a new kernel, the client passes its address and the port of the response socket to the kernel.
- The kernel starts, find free ports to bind the shell, control, stdin, heartbeat and iopub sockets. It then connect to the response socket and sends the connection information to the client.
- Upon reception of the connection information, the client connects to the kernel.

The way the client passes its address and the port of the response socket to the kernel should be similar to that of passing the ports of the kernel sockets in the current implementation: a connection file that can be read by local kernels or sent over the network for remote kernels (although this requires a custom kernel provisioner or "nanny"). This connection file should also contain the signature scheme and the key.
```

Reviewer 2: fwiw, IPython Parallel has a similar socket (because the kernel-client connection direction is reversed) and calls it the "registration socket"

The way the client passes its address and the port of the listening socket to the kernel should be similar to that of passing the ports of the kernel socket in the current implementation: a connection file that can be read by local kernels or sent over the network for distant kernels (although this requires an intermediate actor such as a gateway).

The kernel specifies whether it supports the handshake pattern via the "kernel_protocol_version" field in the kernelspec:
- if the field is missing, or if its value if less than 5.5, the kernel supports passing ports only.
- if the field value is >=5.5 and <6, the kernel supports both mechanisms.
- if the field value is >=6, the kernel supports the handshake pattern. Clients should not assume the kernel still supports the old mechanism.

### Impact on existing implementations

Although this enhancement requires changing all the existing kernels, the impact should be limited. Indeed, most of the kernels are based on the kernel wrapper approach, or on xeus.

Most of the clients are based on `jupyter_client`. Therefore, the changes should be limited to this repository only.

> A transition period where clients and kernels support both mechanisms should allow kernels to gradually migrate to the new version of the protocol. The support of the handshaking pattern should be indicated in the kernelspec.

Author: Good catch, that is an oversight of the previous change.

# Kernel Handshaking pattern

## Problem

The current implementation of Jupyter client makes it responsible for finding available ports and pass them to a new starting kernel. The issue is that a new process can start using one of these ports before the kernel has started, resulting in a ZMQError when the kernel starts. This is even more problematic when spawning a lot of kernels in a short laps of time, because the client may find available ports that have already been assigned to another kernel.

A workaround has been implemented for the latter case, but it does not solve the former one.

## Proposed Enhancement

We propose to implement a handshaking pattern: the client lets the kernel find free ports and communicate them back via a dedicated socket. It then connects to the kernel. More formally:

> - The kernel launcher is responsible for opening a dedicated socket for receiving connection information from kernels (channel ports). This socket will be referred as the **registration socket**.

Reviewer 3: @JohanMabille Would the port registration socket change on each kernel launch, is it static and the socket remains the same for each kernel launch? (in other words, do we have as many registration sockets as kernels, or only one registration socket for all the kernels)?

Author: The two options are theoretically possible (thus the wording "The kernel should not expect the registration socket to exist after it has received the acknowledge receipt (i.e. it can be closed)" in the JEP), although I guess that in most of the implementations we would keep a single socket as long as possible.

Reviewer 3: The spec should be as clear as possible. If both options are possible, it should be stated clearly in words. Further questions:

1. In your **many** registration socket option, what is the added value of this JEP? The initial issue was the availability of the 5 ports. If a registration port needs to be found on each kernel launch, we don't win a lot and are still subject to failure.
2. In your **single** registration socket option, what process is responsible to maintain that socket? Would it be some code in jupyter-client or will each of the "client" (jupyter-server, jupyter-console...) be responsible for that? (I except you answer to be "hey this is an implementation detail...", I hope it will not be)

Author: > The spec should be as clear as possible. If both options are possible, it should be stated clearly in words.

This can be done in the PR to the doc of the protocol  and in the protocol schema. Letting both options opened in the JEP should be enough.

> In your many registration socket option, what is the added value of this JEP? The initial issue was the availability of the 5 ports. If a registration port needs to be found on each kernel launch, we don't win a lot and are still subject to failure.

The current pattern is:
1] the client finds free ports by opening sockets on them
2] the clients *closes* the sockets
3] The clients communicates these ports to the kenrel
3] the kernel tries to open sockets on these ports.

So the problem is that there is a delay between finding free ports and actually opening sockets for using them. With the current pattern, the situation is totally different, even in the "many registration socket option": the kernel finds the free ports by opening sockets on them and then *does not close them*.

> In your single registration socket option, what process is responsible to maintain that socket? Would it be some code in jupyter-client or will each of the "client" (jupyter-server, jupyter-console...) be responsible for that? (I except you answer to be "hey this is an implementation detail...", I hope it will not be)

The process responsible for maintaining that socket is the process that is responsible for launching the kernel, whatever it is, since it is the one passing the registration socket info to the kernel upon launch.

Reviewer 2: I was about to write that registration socket lifetime should not be in the spec, as it's an implementation detail irrelevant to kernels, but then I thought about kernel restarts. Presumably restarting kernels need to re-register their (possibly new) sockets, so the registration socket should stay open (or reopen at the same port - not always available) _IF_ kernel restarts are expected. For example:

> The client MAY close the registration socket after completing a kernel's registration.
> The client MAY use a single registration socket for multiple kernels.
> To restart a kernel will require the registration socket again, so the client SHOULD keep the registration socket open while it expects restarts to be possible, OR open a new socket and pass the new registration socket URL to the new process.

Author: Hey sorry, I missed this one. I will update the JEP accordingly.

- When starting a new kernel, the launcher passes the connection information for this socket to the kernel.
- The kernel starts, find free ports to bind the shell, control, stdin, heartbeat and iopub sockets. It then connect to the registration socket and send the connection information to the client.
- Upon reception of the connection information, the client connects to the kernel.

The way the launcher passes the connection information for the registration socket to the kernel should be similar to that of passing the ports of the kernel socket in the current connection pattern: a connection file that can be read by local kernels or sent over the network for remote kernels (although this requires a custom kernel provisioner or "nanny"). This connection file should also contain the signature scheme and the key.

> The kernel should not expect the registration socket to exist after it has sent its registration (i.e. it can be closed). Therefore, the kernel should disconnect from the registration socket right after it has sent its connection information.

Author: Indeed, I'll amend the JEP.

The kernel specifies whether it supports the handshake pattern via the "kernel_protocol_version" field in the kernelspec:
- if the field is missing, or if its value if less than 5.5, the kernel supports passing ports only.
- if the field value is >=5.5, the kernel supports both mechanisms.

### Remarks

This pattern is **NOT** a replacement for the current connection pattern. It is an additional one and kernels will have to implement both of them to be conformant to the Jupyter Kernel Protocol specification. Which pattern should be used for the connection is decided by the kernel launcher, depending on the information passed in the initial connection file.

> 

Reviewer 3: @JohanMabille  Do we need to specify that coexisting launcher and kernels not supporting both patterns should be possible. e.g. old launcher (single-mode) should be able to launch new kernels (dual-mode)

Similarly, how would.a dual-mode launcher know if the kernel is single- or dual-mode? Is there a "capabilities" JEP for that. I guess it is mandatory for the launcher to know if the kernel support the new handshaking pattern.

Author: As stated in the JEP:

"The kernel specifies whether it supports the handshake pattern via the "kernel_protocol_version" field in the kernelspec:
- if the field is missing, or if its value if less than 5.5, the kernel supports passing ports only.
- if the field value is >=5.5, the kernel supports both mechanisms."

So old launchers can start new kernels, and new launchers will still be able to start old kernels. Reading the kernelspec is enough to know whether a kernel supports both mechanisms or the older one only.

issue-number:
pr-number: 66
date-started: 2021-01-05
---

# Kernel Handshaking pattern

## Problem

The current implementation of Jupyter client makes it responsible for finding available ports and pass them to a new starting kernel. The issue is that a new process can start using one of these ports before the kernel has started, resulting in a ZMQError when the kernel starts. This is even more problematic when spawning a lot of kernels in a short laps of time, because the client may find available ports that have already been assigned to another kernel.

A workaround has been implemented for the latter case, but it does not solve the former one.

## Proposed Enhancement

We propose to implement a handshaking pattern: the client lets the kernel find free ports and communicate them back via a dedicated socket. It then connects to the kernel. More formally:

- The kernel launcher is responsible for opening a dedicated socket for receiving connection information from kernels (channel ports). This socket will be referred as the **registration socket**.
- When starting a new kernel, the launcher passes the connection information for this socket to the kernel.

> - The kernel starts, finds free ports to bind the shell, control, stdin, heartbeat and iopub sockets. It then connects to the registration socket and sends the connection information to the registration socket.

Reviewer 4: ```suggestion
- The kernel starts, finds free ports to bind the shell, control, stdin, heartbeat and iopub sockets. The connection ports information will be referred as the **connection information**. It then connects to the registration socket and sends the connection information to the kernel launcher.
```

Reviewer 4: Should we formally define the socket message format? Or is it irrelevant to a JEP?
